### DISCUSSION TITLE
Videos not playing in waydroid.

### DISCUSSION THREAD

**Discussion by MrDisguised** [waydroid/waydroid#855] [Q&A]
Whenever I play a youtube video or any video in general I see a black screen and the video doesn't play (I can hear the sound playing though). Is there a way to fix it ? I am using archlinux and my wm is hyprland

**Comment by TriMoon** [comment]
I'm having a similar issue as OP, but in my case the videos are initiated inside games which don't show either.
In this case it's adverts provided for ingame benefits that you choose self manually, or videos shown at startup of games.

I'm using Kubuntu 23.04 at moment with nVidia GFX and wayland session, and also have these in `/var/lib/waydroid/waydroid.cfg`:
```
ro.hardware.gralloc = default
ro.hardware.egl = swiftshader
```

**Comment by Seme4eg** [comment]
Exactly same issue when i am using PC with external GPU. Tho on laptop machines with same setup it works fine. The only major difference is that laptop uses internal graphics chip and on PC it uses external GPU. On both machines i use hyprland as compositor and the rest of the software setup is the same.

**Comment by invisiblewizard1** [comment]
Same issue here, playing videos on YouTube or in an app don't produce any errors but the videos won't start playing. Kubuntu 23.04.

**Comment by adocampo** [comment]
Same here. 

Videos are not playing, just sound is heard. 

If I open the browser, load youtube.com, and then changed to "Desktop version", _it seems_ to work. Youtube Shorts are working (at least a few of them, then browser hangs), hovering the mouse over the videos shows the preview (on the mobile version didn't), and normal videos seems to load, but after the image loading, they freeze and finally, all shows this
![image](https://github.com/waydroid/waydroid/assets/2266612/9938808a-3b1b-4861-b45e-37e65fe4a063)

This is on a browser in desktop mode. On a mobile version it behaves like the app, only sound. 

I've tried several video apps:
 - Netflix: is not compatible with this device
 - HBO: no video or sound, subtitles are shown, though
 - Plex: like youtube, only sound. Transcoding doesn't help either. Subs are working
 - Amazon Prime Video: not compatible, it doesn't let install it.
 - Showtime: It crashes and restarts the app when I clic on a video.
 - RTVE: Sound works, video no.
 - Pluto TV: Sound works, video no.

I'm on Arch Linux, Nvidia RTX 3070Ti, usin plasma kwin compositor, based on QtWayland, if that matters.

**Comment by OrlandoRodriguez-KSO** [comment]


**Reply by adocampo** [reply]
how did you install that version? 
I did try to download it manually and extract it on `/usr/share/waydroid-extra/images` and, then did a `sudo waydroid init -f` but nothing happened. If I do a `sudo waydroid init -s GAPPS -f ` it downloads the latest image (as for december) and then, nothing happens again (I mean, all the applications I've installed on my first installation are there, and the issue still is happening)

**Reply by OrlandoRodriguez-KSO** [reply]
Similar to what you have been doing, except by the path (I use /etc/waydroid-extra/images/). I download the system and vendor from the same date, copy and unzip both in those folders and after that, I execute those commands. Visually they are pretty similar, what I have been checking to know if they are different is  "about phone -> Android version".  

![image](https://github.com/waydroid/waydroid/assets/10275325/a7c3bcfa-6537-4f81-9843-2df2bb416345)

This one has been working for me to play video and audio without issues....what is not working is Chrome browser (it did work in newer versions xD in those where the videos don't work). But I just used firefox and that's it.

**Comment by kfirufk** [comment]
any ideas regarding the black screen? i also hear just audio, i tired many apps like youtube, yes+ and others
i'm on gentoo linux ~am64 with latest waydroid with gapps on nvidia-drivers with xanmod kernel with gtx3060.

waydroid is installed using the guru gentoo repository

**Reply by kfirufk** [reply]
I'm not gonna move to xorg just for that, I assume they will fix it cause it's called waydroid not xdroid, I'll just wait for the next release

**Comment by adocampo** [comment]
I couldn't get video working, yet I've tried many older versions, as @orrot said it worked for him an outdated version of lineage, but no luck on my side...

On Arch Linux, there is no ashmem module, as I think it is now in the kernel, binder is a dmks module. I'd installed `binder_linux-dkms`.

**Comment by nischaygarg2006** [comment]
Had the same issue , switching to xorg driver and removing custom config fixed it for me.

**Reply by adocampo** [reply]
could you please tell here how did you do that?

**Reply by adocampo** [reply]
@nischaygarg2006 just tried to start waydroid under X11, then opened a weston window and launched `waydroid show-full-ui` inside of it, but videos are still in black, I think I missed to remove the custom config, but I didn't find any directory under my home... I'm on the good direction?

**Comment by OrlandoRodriguez-KSO** [comment]
Hey @adocampo  I just formated and install a new distro. I installed Fedora 39 in my laptop that previously had ubuntu, and I just installed waydroid and everything worked really well, less steps and no troubleshooting to get Waydroid with Udemy and youtube together. I am wondering if the problems with waydroid might not be related to the Kernel or something like that. What distro are you using ?

**Reply by adocampo** [reply]
I'm on Arch Linux, kernel 6.7.0 now, but I've been using waydroid since 6.4 or so.
```
Operating System: Arch Linux 
Kernel Version: 6.7.0-arch3-1 (64-bit)
Processors: 16 × 11th Gen Intel® Core™ i7-11700K @ 3.60GHz
Memory: 46.9 GiB of RAM
Graphics Processor: NVIDIA GeForce RTX 3070 Ti/PCIe/SSE2
```

**Comment by YAFU** [comment]
Hi.
Same problem with Waydroid related to black screen, no video and only audio (PlutoTV, YouTube, etc). KDE Neon User Edition, nvidia GTX 960 proprietary driver.

So this is mainly related to nvidia proprietary driver?
Do the developers have any clues on where to look/investigate/experiment about the cause of the problem?

**Comment by B4rabbas** [comment]
Same, the error occur even with a drm free mp4 file.

**Comment by eagleivg** [comment]
The problem seems to be specific to chromium, the video plays fine in Fennec browser from F-Droid

**Reply by adocampo** [reply]
F-Droid says it is not compatible
![image](https://github.com/waydroid/waydroid/assets/2266612/ab29a78a-7a39-4614-9afc-0d8fdc450c25)

**Reply by eagleivg** [reply]
Oh, im use waydroid in ARM64 platform. Тry get x86_64 build of fennec: https://github.com/mozilla-mobile/firefox-android/releases/download/fenix-v123.1.0/fenix-123.1.0-x86_64.apk

**Reply by adocampo** [reply]
Perhaps in ARM64 videos are working, but on x86_64 aren't working, at least with NVIDIA cards, or perhaps is because I'm on arch linux. I've tried version 123 and even 124 of firefox android with the same result: audio plays, videos are black.

**Comment by davidbitterlich** [comment]
Hi, I have the very same issue using Ubuntu 22.04. Hope there will be a solution for this issue. I specifically wanted to test running Waydroid with streaming services, such as Prime Video and Disney+ with Widevine integrated...

**Reply by eagleivg** [reply]
What sound server are you using, pipewire or pulseaudio?

**Reply by davidbitterlich** [reply]
I use PulseAudio, that's the default on Ubuntu 22.04

**Comment by AlanPidal** [comment]
I've got the same issue and tried tonsolve by forcing Waydroid to run in VM mode but didn't work. I'm currently using Ubuntu 22.04 with Wayland GNOME. I tried downloading the videos AND playing them un the gallery but It also failed  I've also tried the method of installing oldee Versions of Lineage OS but didn't work. Kinda surprised by people saying this suddenly happened. Hope It gets solved :/ but not sure what to do.

**Comment by daniil7** [comment]
Kubuntu 22.04. the same problem with the fresh installation of Waydroid. Any application does not respond when watching a video or shows a black screen or a static picture. After installing this ubuntu-restricted-extras package, the video works in PipePipe (youtube frontend) and locally in VLC media player, but not in browsers, I tried the standard browser and duckduckgo.
I do not understand what happened, and whether the installation of ubuntu-restricted-extras affected the fact that video playback now works in non-browser applications.

**Comment by eagleivg** [comment]
So, after a few experiments, the following was found out:
It's a bug in the PCM implementation, it appears in both pulseaudio and pipewire-pulse. 
I managed to get normal audio playback in video at the same time by throwing /dev/snd into the waydroid container and specifying hardware PCM for SND_PCM_STREAM_PLAYBACK

In my case, this:
hardware/waydroid/audio/audio_hw.c
```
-        ret = snd_pcm_open(&out->pcm, "pulse", SND_PCM_STREAM_PLAYBACK, 0);
+        ret = snd_pcm_open(&out->pcm, "hw:0,2", SND_PCM_STREAM_PLAYBACK, 0);
```

But, its not generic solution, need additional research of pulse playback

**Reply by eagleivg** [reply]
The issue is that the waydroid implementation of sound HAL contains a get_presentation_position callback, which is used to determine the current playback position. And it works correctly for hw ALSA devices!

But everything changes when pulseaudio comes in - now, it returning 0 there. This is the reason why programs using the Android WebView component have this problem. The easiest solution is to remove this callback - then the position calculation is done inside WebView and the video plays correctly. Although the most correct solution would be to fix the snd_pcm_htimestamp call for pulse PCMs, but it is longer and more difficult.

I'll make a PR in the near future.

**Comment by eagleivg** [comment]
Made the PR: https://github.com/waydroid/android_hardware_waydroid/pull/55
Please, test this, solve yours bug or not? It fixed for me, and yes, bug affected at enable video HW acceleration

**Comment by adocampo** [comment]
I would love to test it, but as an end user, I don't know how can I do it... :|

**Reply by eagleivg** [reply]
[audio.primary.waydroid.so.zip](https://github.com/waydroid/waydroid/files/14921401/audio.primary.waydroid.so.zip)
Just upgrade your waydroid installation. Or you can manyally override library via overlayfs (unzip and copy .so library), in my case its /var/lib/waydroid/overlay/vendor/lib/hw/audio.primary.waydroid.so at OpenMandriva

**Comment by sbcdave** [comment]
Did this ever get fixed and merged into Master? I'm on lineage_waydroid_x86_64-userdebug 11 RQ3A.211001.001 89 test-keys
18.1-20240504-GAPPS-waydroid_x86_64 and don't seem to be able to watch any video in waydroid. No YouTube or browser videos. My primary hope was to get the Owlet baby monitor camera app available on one of my monitors while I work. I have waydroid and Owlet running, and I can hear, but the video is just black. /var/lib/waydroid/overlay/vendor is empty

**Reply by eagleivg** [reply]
My fix was merged:  https://github.com/waydroid/android_hardware_waydroid/commit/c93b38970442d0f558721198eea9a0258d8f0aa3
So, if you use latest version (older than 25-th Marth), it can be already fixed, and no need override any library for this.

You can try next hack: find library audio.primary.default.so, copy it to /var/lib/waydroid/overlay/vendor/lib/hw/ and rename it  to audio.primary.waydroid.so
So, if video playback will be fixed (with no sound), than problem in the sound subsystem. If not - someone else.

**Reply by sbcdave** [reply]
Thank you. My waydroid vs was 1.4.2 from October 2023 (via apt install), so I apt removed it, cloned from master, make installed, and tried again, but still have same issue. I also tried building lib/hw inside /var/lib/waydroid/overlay/vendor, and then copied the audio.primary.waydroid.so from you're previous post into there, stopped container, started container, started session, waydroid show-full-ui, still no luck. You got any other ideas?

**Reply by sbcdave** [reply]
Just noticed since building the new version and using it...when I fast forward through a youtube video I can see the stills now. Also noticed that if I put your audio.primary.waydroid.so in /var/lib/waydroid/overlad/vendor/lib/hw/. the videos show as green instead of black and audio stops working.

**Reply by eagleivg** [reply]
Based on the symptoms listed above - the problem described does not sound like a sound problem. Could it be a hardware acceleration problem?

**Reply by TeeCker** [reply]
Thank you for your advices but I was trying everything descriebed above, but with no success. I tried it on nvidia card (proprietary nvidia-driver-470 and nouveau) and also on intel IG (driver=i915), but with the same results. Without the PATCH "/var/lib/waydroid/overlay/vendor/lib/hw/audio.primary.waydroid.so", audio is working but video is black and with the PATCH, the videos show as green instead of black and audio is not working at all...Please any clues ? Thank you in advance :-)

**Reply by TeeCker** [reply]
> You can try next hack: find library audio.primary.default.so, copy it to /var/lib/waydroid/overlay/vendor/lib/hw/ and rename it to audio.primary.waydroid.so
So, if video playback will be fixed (with no sound), than problem in the sound subsystem. If not - someone else.

with this hack i got the same results as with using your audio.primary.waydroid.so library: YouTube App: the videos show as green instead of black and audio is not working at all.

**Reply by thaynes43** [reply]
Just did a fresh install of Waydroid in an Ubuntu VM and I am also experiencing black screen for all video but scrubbing shows picture and audio comes through fine. I followed the steps here https://docs.waydro.id/usage/install-on-desktops#ubuntu-debian-and-derivatives. Not sure if I have the version w/ the fix though, this is all I see:

![image](https://github.com/waydroid/waydroid/assets/4860853/6ec215d3-bd58-433b-a792-b786040a6556)

On the ubuntu side videos play fine.

**Reply by sbcdave** [reply]
Scrubbing? "On the Ubuntu side videos play fine"?

**Reply by thaynes43** [reply]
> Scrubbing? "On the Ubuntu side videos play fine"?

Sorry, I'm not much of a youtuber but my kids are... I mean when you move the timestamp the video is at with the bar below the video, it will show a preview of what the video is at the time you move it to and like a graph of how popular over time the sections of the video are. That works fine but normal playback is a black screen.

And for the ubuntu side, I have waydroid installed on an Ubuntu VM. If I use chrome or firefox on that VM and go to YouTube or Plex videos can play just fine. When I use the apps installed by google play store they are all black screens w/ only sound.

**Reply by sbcdave** [reply]
Gotcha thanks. I'll see if loading in a browser in a vm works for me

**Comment by Timbo303** [comment]
This might not be a fix for everyone but I had this issue when I was upgrading Kubuntu to 24.04 from 23.10. This solution I found requires smart dock.

1. Right Click the app you want to open in all apps.
2. Open in Maximized.

Somehow it fixed it for me. Might not work though for everyone. Might had been something else entirely which is why I included the logcat. Good luck its a long file.

[waydroid-logcat.txt](https://github.com/user-attachments/files/15515790/waydroid-logcat.txt)

**Comment by terodactilo21** [comment]
I'm experiencing the same issue:
-Fedora 40 KDE Plasma 6.1 with Wayland (no X11). 
- GPU: NVIDIA RTX 3090 (w 560 driver)
- Waydroid 1.4.3

I've installed Waydroid and it's running, but when trying to play videos, I only see a black screen. This includes YouTube and any other video playback attempt. The audio works fine, but there's no video display.

I've already installed the necessary codecs and RPM Fusion prior to installing Waydroid. All movies and videos play correctly natively on my PC outside of Waydroid.

I've tried several troubleshooting steps, including:

1. Checking and attempting to change the ro.hardware.egl property (currently set to swiftshader)
2. Ensuring all codecs are installed
3. Updating Waydroid to the latest version
4. Checking logs for specific errors

None of these steps have resolved the issue. I've also read about potential solutions involving modifying the PCM implementation or audio_hw.c file, but these seem to require significant technical intervention and aren't official solutions.

This problem appears to be persistent across different Linux distributions and hardware configurations, suggesting it might be a more fundamental issue with Waydroid's video handling, especially on systems using NVIDIA GPUs with Wayland.

Any official guidance or potential solutions would be greatly appreciated, as this significantly impacts the usability of Android apps requiring video playback within Waydroid.

**Reply by DotGabe** [reply]
Same issue guys.
Rtx 3060
Fedora 41 KDE with wayland.

I've tried modifying audio_hw file, smartdock with maximized window, and I've tried spoof too (https://github.com/waydroid/waydroid/issues/1596#issuecomment-2430606127) and nothing worked.

**What did you solve?**
I'm installed custom image ([Waydroid-ATV](https://github.com/supechicken/waydroid-androidtv-build/releases/tag/20250327)) and worked pretty well with apps tested (Mubi and youtube)

**Comment by ZReC** [comment]
I'm having a rather peculiar glitch with video playback using swiftshader (nvidia gpu). Basically, the video's aspect ratio is vertically cut in half, rendering a 16:9 video in 16:4.5, effectively shrinking it. Should I open an issue?

**Comment by aleasto** [comment]
Video playback with software acceleration (such as when using NVIDIA hardware) will be fixed in Android 13

**Comment by rastla** [comment]
I have the same issue on Nobara 42 with KDE Wayland and AMD Ryzen 5800X and AMD RX 6900 XT
YouTube video playback mostly works (until I press F11 to get it into windowed mode, then it breaks).
Video playback of mp4 file in gallery also works
But e.g. Instagram video playback does not work. Audio still works, but video stays black

Tried deactivating software acceleration, didn't make a difference

**Comment by UpFourAGoodTime** [comment]
Do any of you have ARM translation libraries, like libhoudini?

**Comment by lazyBoy112** [comment]
For those experiencing the same issue as me on Ubuntu 22.04 using PipeWire, where no audio plays and videos won't start: try updating your PipeWire packages to a newer version
